### DISCUSSION TITLE
Should we use enable, disable or annotations nullable context for for unit tests?

### DISCUSSION THREAD

**Discussion by codequality44** [xunit/xunit#2541] [Question] [answered]
My question is a little bit aside from xUnit, but I believe each xUnit user, who is also using Asp.net Core 6 has this pain.

I have a Asp.net Core 6 web api application with xUnit unit and integration tests.

The `<Nullable>enable</Nullable>` setting is automatically added when you add xUnit test library. On the code project - it is ok, but on the Unit test and Integration test projects, as you can imagine - I get compiler warnings for possible null reference exception for things like this:

`        viewModel.Message.Should().Be("Aaaa");
`
So I will need to put the "!" (damnit) operator after every viewModel - to tell the compiler - I am sure the viewModel is not null in this test - because this is how I designed the test! 

I am wondering what the most appropriate nullability context for tests is?

I am currently leaning towards <Nullable>annotations</Nullable>, because it will allow me to use the "?" operator - I have parameters which I pass deliberately as null, and I will not have to use "!" all the time.

What do you think - what is the best context for unit tests?

**Comment by bradwilson** [comment, marked as answer]
There are two answers.

First: You're using some 3rd party assertion system (the `.Should().Be()` stuff), which is outside our control. We cannot add nullable annotations to that.

Second: For the assertion library that we ship, as of 2.4.2 we include include nullable annotations, so things like `Assert.NotNull(x);` will inform the parser that `x` will not be null after that call.
